python-qcgrids/qcgrids/ext.pyx

```diff
@@ -157,3 +167,386 @@ cdef class Cellgrid(object):
     def integrate(self, *factors):
         subscripts = '%s->' % (','.join(['i']*(len(factors)+1)))
         return np.einsum(subscripts, self.weights, *factors)
+
+
+cdef class ScalarFunction(object):
+    """Abstract base class for all scalar functions."""
+
+    def __init__(self):
+        raise RuntimeError("Cannot instantiate abstract base class ScalarFunction.")
+
+    property invertible:
+        def __get__(self):
+            """Is the function invertible or not?"""
+            return deref(self._this).invertible()
+
+    def calc(self, xpy, int nderiv):
+        """Compute the function value and optional first and second derivatives.
+
+        Parameters
+        ----------
+        xpy : float or np.ndarray
+            The input to the function.
+        nderiv : int
+            The number of derivatives (0, 1, 2).
+
+        Returns
+        -------
+        output : np.ndarray
+            Array with ``shape = xpy.shape + (nderiv + 1,)``. The last index is used for
+            function value (0), first derivative (1) and second derivative (3).
+        """
+        x = np.asarray(xpy, dtype=float)
+        output = np.zeros(x.shape + (nderiv + 1,), dtype=float)
+        cdef np.flatiter itx = np.PyArray_IterNew(x)
+        cdef int xndim = x.ndim
+        cdef np.flatiter ito = np.PyArray_IterAllButAxis(output, &(xndim))
+        while np.PyArray_ITER_NOTDONE(itx):
+            deref(self._this).calc(
```

Comment:
Is there any way to avoid stuff like this? It's worse than C++!

Reply:
Do you mean the Numpy C iterator API for getting ufunc-like behavior? (See https://docs.scipy.org/doc/numpy-1.13.0/reference/c-api.iterator.html) I don't think their is a more C++-ish counterpart for it.

I believe `deref(self._this)` is the cleaner and formally correct way to dereference a C++ pointer in Cython. See http://cython.readthedocs.io/en/latest/src/userguide/wrapping_CPlusPlus.html#c-operators-not-compatible-with-python-syntax The more common alternative is `self._this[0]`, which is not realy nice either and strictly speaking not a dereference.

I agree that it is worse than C++, it Cython. :)